crypto/src/lib.rs

```diff
@@ -186,28 +136,64 @@ ffi::ffi_item! {
     }
 }
 
+#[cfg(feature = "rand")]
 impl KeyPair {
-    /// Generates a pair of Public and Private key with [`Algorithm::default()`] selected as generation algorithm.
-    #[cfg(feature = "rand")]
-    pub fn generate() -> Self {
-        Self::generate_with_configuration(
-            KeyGenConfiguration::from_random().with_algorithm(Algorithm::Ed25519),
+    /// Generate a random key pair using a default [`Algorithm`].
+    pub fn random() -> Self {
+        Self::random_with_algorithm(Algorithm::default())
+    }
+
+    /// Generate a random key pair
+    pub fn random_with_algorithm(algorithm: Algorithm) -> Self {
+        macro_rules! with_algorithm_variations {
+            ($(($alg:ident, $alg_mod:path)),+) => {
+                match algorithm {
+                    $(Algorithm::$alg => <$alg_mod>::keypair(KeyGenOption::Random).into()),*
+                }
+            }
+        }
+
+        with_algorithm_variations!(
+            (Ed25519, ed25519::Ed25519Sha512),
+            (Secp256k1, secp256k1::EcdsaSecp256k1Sha256),
+            (BlsNormal, bls::BlsNormal),
+            (BlsSmall, bls::BlsSmall)
         )
     }
 }
 
 #[ffi_impl_opaque]
 impl KeyPair {
+    /// Derive a key pair from a seed using pRNG
+    pub fn from_seed(seed: Vec<u8>, algorithm: Algorithm) -> Self {
+        macro_rules! with_algorithm_variations {
+            ($(($alg:ident, $alg_mod:path)),+) => {
+                match algorithm {
+                    $(Algorithm::$alg => <$alg_mod>::keypair(KeyGenOption::UseSeed(seed)).into()),*
+                }
+            }
+        }
+
+        with_algorithm_variations!(
+            (Ed25519, ed25519::Ed25519Sha512),
+            (Secp256k1, secp256k1::EcdsaSecp256k1Sha256),
+            (BlsNormal, bls::BlsNormal),
+            (BlsSmall, bls::BlsSmall)
+        )
+    }
+
     /// Algorithm
     pub fn algorithm(&self) -> Algorithm {
         self.private_key.algorithm()
     }
 
-    /// Construct a [`KeyPair`]
+    /// Construct a [`KeyPair`].
+    ///
+    /// See [`Self::into_raw_parts`] for an opposite conversion.
     ///
     /// # Errors
-    /// If public and private key don't match, i.e. if they don't make a pair
-    pub fn new(public_key: PublicKey, private_key: PrivateKey) -> Result<Self, Error> {
+    /// If public and private keys don't match, i.e. if they don't make a pair
+    pub fn from_raw_parts(public_key: PublicKey, private_key: PrivateKey) -> Result<Self, Error> {
```

Comment:
See https://github.com/hyperledger/iroha/issues/4339#issuecomment-1984711378

I finally renamed it to `from_raw_parts`, because there is a mirrored method called `into_raw_parts` and it makes sense.

Reply:
the reason why we need a destructuring method is because we want to keep fields of our structs private but also not be required to clone if we want to move a field out of a struct. This is the same in `iroha_crypto` and `iroha_data_model` so we better find a general solution to this. And having constructors named `from_raw_parts` is not what I would like


IMO `raw_parts` implies some sort of raw pointers when you look at `std`. Instead of `KeyPair::into_raw_parts` we already have:
```rs
impl From<KeyPair> for (PublicKey, PrivateKey)
```

so we can just remove `KeyPair::into_raw_parts`. We can also just remove `KeyPair::new` and add:
```rs
impl From<(PublicKey, PrivateKey)> for KeyPair
```

Reply:
let's have a discussion on this at standup

Reply:
- let's keep `KeyPair::new`
- rename `KeyPair::into_raw_parts` into `KeyPair::into_parts`
- remove `From<KeyPair> for (Pk, Pk)`

Reply:
Addressed in [`ec0fa3a` (#4341)](https://github.com/hyperledger/iroha/pull/4341/commits/ec0fa3a1a4185f11deabb2cffa5c895a50f7569e), and created https://github.com/hyperledger/iroha/issues/4360 as discussed